EIPS/eip-draft_disable_selfdestruct.md

```diff
@@ -0,0 +1,63 @@
+---
+eip: <to be assigned>
+title: Disable SELFDESTRUCT
+author: Alexey Akhunov (@AlexeyAkhunov)
+discussions-to: https://ethereum-magicians.org/t/eip-for-disabling-selfdestruct-opcode/4382
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-06-25
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+Disable SELFDESTRUCT
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+Make `SELFDESTRUCT` EVM operation no-op (effectively disable). Contracts will not be able to delete themselves.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+Althouth `SELFDESTRUCT` originally came with EVM to help clean up the state, we learnt that in practice it did not achieve this objective on
+sufficient scale, and is on balance a burden of extra complexity and unintended effects (GasToken2 and polymorphic contracts).
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+Operation `SELFDESTRUCT` (formerly `SUICIDE`) was in the EVM from the beginning, and its purpose was to incentivise clearing the state, by giving the
+caller gas refund. In practice, this incentivisation turned out to be very limited, and the designed purpose was not archieved. However, `SELFDESTRUCT`
+brings significant amount of complexity to the EVM. It is responsible for some of the most arcane edge cases in the EVM semantics.
```

Comment:
Strongly agree on this. In particular, the existence of this opcode significantly increases the complexity of many forms of caching, as SELFDESTRUCT can potentially delete an unbounded number of storage slots.

Reply:
If selfdestruct did not delete those storage slots it would empower a new kind of upgrade pattern (create2 reincarnations) while reducing SELFDESTRUCT complexity.

I don't think geth ever clears those storage slots from the state trie; it checks if there was a recent self-destruct. A reincarnation could fork the network if not enough nodes remain aware of the selfdestruct and reference a quantum storage slot.